HW0/P5/P5.txt

```diff
@@ -1 +1,146 @@
 # Your answers here.
+
+Part 1. (Answer)
+Suppose we have N bags. 
+It takes 1 second to add up the first two bags (so there are N-2 bags left);
+thereafter it takes 1 second to add one more bag to the running sum. 
+So it takes N-1 seconds for one person to add up all the bags.
+
+*********************************
+In this case, it takes 255 seconds for 256 bags.
+*********************************
+
+Part 2. (Answer)
+Assuming one person can only add two numbers per second. 
+Assume that for any given sum, computation of the two summands are independent
+tasks (that is, knowing one summand of the sum doesn't speed up the computation
+of the other and that computing one summand doesn't require the knowledge of the 
+other).
+
+The person who computed the total, at time t_f, could have added up only two 
+subtotals; so, given our assumption, the best case scenario is that both subtotals 
+were produced optimally and simultaneously (so t_f - 1 is the min of the optimal 
+compute times of both subtotals).
+
+For M number of subtotal computed at time t > 0, the same rationale applies, that is, 
+we want to have optimally and simultaneously computed the 2M summands by time t-1. Thus, 
+at time 1 we want to calculate floor(N/2) number of subtotals; at time t we want to 
+calculate floor(N/2^t) number of subtotals.
+
+If we had an employee ready to perform a calculation whenever one was needed, 
+the fastest we can compute the total is ceil(log_2(N)).
+
+In our case, at best, we can compute the total for 256 bags in 8 seconds.
+
+Note that at each time t, we need floor(N/2^t) number of employees to do simultaneous
+calculations. Since the number of calculations (hence employees) is monotone decreasing
+with respect to time, the bottle neck is the number of employees needed at time t = 1.
+So, if the number of employees, M, is less than floor(N/2). Then at each time t, we can 
+reduce the number of subtotals that remain to be summed by at most M. Let x be the least 
+natural number such that M - xN >= 2M. Then the fastest we can compute the total will be 
+x + ceil(log_2(M-xN)).
+
+*********************************
+In our case, we can compute the total for 256 bags in (30 + 4) seconds with 8 employees.
+*********************************
+
+Part 3. (Answer)
+[See answer to Part 2]
+
+Part 4. (Answer)
```

Comment:
(Adding my comments here since I can't directly comment on the graph.) To graph the difference between the n and log_2(n), I made the y axis logarithmic. This would make the difference between the two clear for very large n.